### ISSUE TITLE
Cache javascript functions in-browser

### ISSUE THREAD

**Issue by sanity** [kwebio/kweb-core#53]
Currently Kweb sends instructions to the browser, which are either predefined for common operations (eg. `createElement`), or executed using [eval()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval) for rarer operations.  These operations can be triggered using [execute()](https://github.com/kwebio/kweb-core/blob/a5a68684fdce1f2326a25514cc7bd48c17b86ba1/src/main/kotlin/io/kweb/WebBrowser.kt#L48), [evaluate()](https://github.com/kwebio/kweb-core/blob/master/src/main/kotlin/io/kweb/WebBrowser.kt#L60), and some other related functions on the [WebBrowser](https://github.com/kwebio/kweb-core/blob/master/src/main/kotlin/io/kweb/WebBrowser.kt#L60) class.

This approach should be replaced by one in which the first call to a JavaScript function results in the function being sent to the client and parsed using [new Function()](https://davidwalsh.name/new-function).  

This could be exposed by modifying `WebBrowser.execute()` and related functions to support, perhaps something like:

```kotlin
   Kweb(port = ..) { 
     val username : String = // ..
     execute("""alert("Hi {name}, I'm an alert box");""", "name" to username)
   }
```
Importantly, this function is then cached such that future calls to the function (with different parameters) can be sent to the browser with only a reference to the cached function and the parameter values.

The advantages here are to remove any `eval()` overhead, while replacing the current Instruction mechanism with something that achieves the same efficiency benefit but without having to manually create each special case.

An implementation thought: We should probably use randomly generated variable names for the parameters as a precaution against variable name clashes, unless there is a way to ensure that can't happen.

If we go with that syntax for the function template we can use Ktor's RoutingPath, as we already do in [routing.kt](https://github.com/kwebio/kweb-core/blob/master/src/main/kotlin/io/kweb/routing/routing.kt).

**Comment by sanity** [comment]
This is going to be a fairly major rework of the server-browser communications, it should probably combined with an implementation of SSR - rendering the initial DOM state to the HTML page before it is sent to the user.  This should contribute significantly to SEO and perhaps to page rendering performance.

**Comment by sanity** [comment]
> This is going to be a fairly major rework of the server-browser communications, it should probably combined with an implementation of SSR - rendering the initial DOM state to the HTML page before it is sent to the user.  This should contribute significantly to SEO and perhaps to page rendering performance.

Note: SSR is now implemented.

**Comment by sanity** [comment]
Enhancement merged to master from pull request #180, will be in next release.
